sdk/tests/conformance/misc/views-with-offsets.html

```diff
@@ -0,0 +1,353 @@
+** TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+** MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+*/
+
+-->
+
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<link rel="stylesheet" href="../../resources/js-test-style.css"/>
+<script src="../../js/js-test-pre.js"></script>
+<script src="../../js/desktop-gl-constants.js"></script>
+<script src="../../js/webgl-test-utils.js"></script>
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+
+<script>
+"use strict";
+description("Tests texture uploads with ArrayBufferView+offsets");
+
+var wtu = WebGLTestUtils;
+var contextVersion = wtu.getDefault3DContextVersion();
+var gl = wtu.create3DContext();
+console.log(gl.getParameter(gl.VERSION));
+
+////
+
+function arrToStr(arr) {
+  return "[" + arr.map(x => x.toString()).join(", ") + "]";
+}
+
+function shouldBeWas(shouldBe, was, info) {
+  var text = "Should be " + shouldBe + ", was " + was + ".";
+  if (info) {
+    text = info + ": " + text;
+  }
+
+  if (shouldBe == was) {
+    testPassed(text);
+    return true;
+  } else {
+    testFailed(text);
+    return false;
+  }
+}
+
+function shouldBeWasArr(shouldBe, was, info) {
+  if (shouldBe.length != was.length) {
+    testFailed("Length should be " + shouldBe.length + ", was " + was.length + ".");
+    return false;
+  }
+
+  return shouldBeWas(arrToStr(shouldBe), arrToStr(was), info);
+}
+
+////
+
+// Textures
+
+var fibArr = [
+  0, 1, 1, 2,
+  3, 5, 8, 13,
+  21, 34, 55, 89,
+  144, 233,
+];
+
+var fb = gl.createFramebuffer();
+
+function probeWithBadOffset(fnTest, info) {
+  fnTest(+(-1|0));
+  if (!gl.getError()) {
+    var text = "Does not support " + info + " with offsets into views.";
+    if (contextVersion >= 2) {
+      testFailed(text);
+    } else {
+      testPassed("Warning: " + text);
+    }
+    return false;
+  }
+  return true;
+}
+
+// fn(view, offset, expectedError, expectedResult)
+
+do {
+  var readPixelView = new Uint8Array(4);
+  var testView = new Uint8Array(fibArr);
+
+  function testTexOrSubImage(funcName, fnTexOrSubImage) {
+    debug("");
+    debug(funcName);
+
+    var fnProbe = function(viewOffset) {
+      fnTexOrSubImage(gl.RGBA, gl.UNSIGNED_BYTE, testView, viewOffset);
+    };
+
+    if (!probeWithBadOffset(fnProbe, funcName))
+      return;
+
+    for (var i = 0; i <= testView.length+1; i++) {
+      debug("offset=" + i);
+
+      fnTexOrSubImage(gl.RGBA, gl.UNSIGNED_BYTE, testView, i);
+
+      var effectiveViewLen = testView.length - i;
+
+      if (effectiveViewLen >= 4) {
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+        gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, readPixelView);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+        shouldBeWasArr(testView.slice(i, i+4), readPixelView);
+
+      } else if (effectiveViewLen >= 0) {
+        wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION);
+
+      } else {
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
```

Comment:
The "srcOffset" argument to texImage2D / texSubImage2D taking ArrayBufferView is GLuint, so INVALID_VALUE can't be generated for negative arguments; they look like large positive arguments, and generate INVALID_OPERATION.